github/github_test.go

```diff
@@ -369,64 +261,149 @@ func TestHandlePullRequestCreateLabelError(t *testing.T) {
-	mockRepositoryCommits := []*github.RepositoryCommit{{Author: &github.User{Login: &mockAuthorLogin}}}
-	forcedError := fmt.Errorf("forced AddLabelsToIssue error")
-	GHImpl = &GHInterfaceMock{
-		PullRequestsMock: PullRequestsMock{mockRepositoryCommits: mockRepositoryCommits},
-		IssuesMock: IssuesMock{
-			mockGetLabel:       &github.Label{},
-			mockAddLabelsError: forcedError,
+		err := HandlePullRequest(logger, mockDB, prEvent, 0, "")
+		assert.EqualError(t, err, forcedError.Error())
+	})
+
+	t.Run("TestHandlePullRequestIsCollaboratorTrueCollaborator", func(t *testing.T) {
+		authors := []string{"anAuthor4"}
+		issuesMock := IssuesMock{
 			MockGetLabelResponse: &github.Response{
 				Response: &http.Response{},
 			},
-		},
-	}
+			MockRemoveLabelResponse: &github.Response{
+				Response: &http.Response{},
+			},
+		}
+		repositoriesMock := RepositoriesMock{
+			isCollaboratorResult: true,
+		}
+		GHImpl = getGHMock(getMockRepositoryCommits(authors, true), &issuesMock, &repositoriesMock)
 
-	prEvent := webhook.PullRequestPayload{}
+		prEvent := webhook.PullRequestPayload{}
 
-	mockDB, logger := setupMockDB(t, true)
-	mockDB.hasAuthorSignedLogin = mockAuthorLogin
+		mockDB, logger := setupMockDB(t, true)
+		mockDB.hasAuthorSignedLogin = authors[0]
+		mockDB.removePRsEvalInfo = &types.EvaluationInfo{}
 
-	err := HandlePullRequest(logger, mockDB, prEvent, 0, "")
-	assert.EqualError(t, err, forcedError.Error())
+		err := HandlePullRequest(logger, mockDB, prEvent, 0, "")
+		assert.NoError(t, err)
+	})
+
+	t.Run("TestHandlePullRequestListCommitsError", func(t *testing.T) {
+		forcedError := fmt.Errorf("forced ListCommits error")
+		GHImpl = &GHInterfaceMock{
+			RepositoriesMock: *setupMockRepositoriesService(t, false),
+			PullRequestsMock: PullRequestsMock{
+				mockListCommitsError: forcedError,
+			},
+		}
+
+		// GHImpl = getGHMock(nil, nil, setupMockRepositoriesService(t, false))
+
+		prEvent := webhook.PullRequestPayload{}
+		mockDB, logger := setupMockDB(t, true)
+		err := HandlePullRequest(logger, mockDB, prEvent, 0, "")
+		assert.EqualError(t, err, forcedError.Error())
+	})
+
+	t.Run("TestHandlePullRequestListCommits", func(t *testing.T) {
+		mockExternalUrl := "fakeExternalURL"
+		GHJWTImpl = &GHJWTMock{
+			AppsMock: AppsMock{
+				mockInstallation: &github.Installation{
+					AppSlug: &appSlug,
+				},
+				mockAppResp: &github.Response{Response: &http.Response{StatusCode: http.StatusOK}},
+				mockApp:     &github.App{ExternalURL: &mockExternalUrl},
+			},
+		}
+
+		origGithubImpl := GHImpl
+		defer func() {
+			GHImpl = origGithubImpl
+		}()
+		authors := []string{"john", "doe"}
+		GHImpl = getGHMock(getMockRepositoryCommits(authors, true), nil, nil)
+		prEvent := webhook.PullRequestPayload{}
+		mockDB, logger := setupMockDB(t, false)
+		err := HandlePullRequest(logger, mockDB, prEvent, 0, "")
+		assert.NoError(t, err)
+	})
+
+	t.Run("TestHandlePullRequestListCommitsUnsignedCommit", func(t *testing.T) {
```

Comment:
Great to see the unit test!!!

Scratching my head to figure out a "nice"(tm) way to assert the code behaved the way we wanted (besides returning no error). Not a blocker in my opinion. Just a code smell for us to consider going forward. Might be an indication the method `EvaluatePullRequest()` is growing too large, and/or we need to be able to test smaller "units" - but even then the question of which side effects or return values to assert remains...

Nicely done! Super helpful user error message too.

Reply:
I agree with `EvaluatePullRequest()` is too large...